File: cpp/leetcode/008.cpp
class Solution {
public:
    int myAtoi(string str) {
        if (str.size() == 0) {
            return 0;
        }
        int sign = 1, i = 0;
        long long res = 0;
        for (; i < str.size() && str[i] == ' '; i++);
        if (str[i] == '+' || str[i] == '-') {
            sign = str[i++] == '+' ? 1 : -1
        }
        while (i < str.size()) {
            if (str[i] >= '0' && str[i] <= '9') {
                res = 10 * res + str[i++] - '0';
            } else {
                break;
            }
            if (res * sign > INT_MAX) {
                return INT_MAX;
            } else if (res * sign < INT_MIN) {
                return INT_MIN;
            }

        }
        return (int) res * sign;
    }
};

File: python/leetcode/008.py
class Solution(object):
    def myAtoi(self, str):
        str = str.strip()
        if not str:
            return 0
        INT_MAX, INT_MIN = 2**31 - 1, -2**31
        sign, i, res = 1, 0, 0
        if str[0] in "+-":
            i += 1
            sign = 1 if str[0] == '+' else -1
        while i < len(str):
            if str[i].isdigit():
                res = 10 * res + int(str[i])
                if sign * res > INT_MAX:
                    return INT_MAX
                if sign * res < INT_MIN:
                    return INT_MIN
                i += 1
            else:
                return sign * res
        return sign * res
